src/content/blog/ddev-and-docker-healthchecks-technote.md

```diff
@@ -0,0 +1,93 @@
+---
+title: "DDEV and Docker Healthchecks (Tech Note)"
+pubDate: 2024-07-22
+# modifiedDate: 2024-07-23
+summary: "DDEV and Docker Healthchecks (Tech Note)"
+author: Randy Fay
+featureImage:
+  src: /img/blog/2024/07/technote.png
+  alt: "DDEV and Docker Healthchecks (Tech Note)"
+  credit: 'Dall-E: A banner graphic representing a "technical note" concept. The design should feature a paper background suggestive of a technical document'
+categories:
+  - TechNotes
+---
+
+## What are Docker Container Healthchecks?
+
+Docker provides [Container Health Checks](https://docs.docker.com/reference/dockerfile/#healthcheck) as a means of verifying whether a container is behaving correctly, whether its services or configuration are working right. Normally a healthcheck is a script selected via the Dockerfile configuration or a [Docker Compose specification](https://github.com/compose-spec/compose-spec/blob/main/spec.md#healthcheck).
+
+## How does DDEV use Healthchecks?
+
+DDEV's primary use of healthchecks is to make sure that everything is working inside a container before declaring it ready for use (and continuing startup). For example, during `ddev start` DDEV starts `ddev-webserver` and `ddev-dbserver` by issuing a `docker-compose up`. It then waits for `healthy` status on each of these. For `ddev-webserver`, this will mean that the web server is properly able to serve files, that `php-fpm` or `gunicorn` is able to interpret scripts, and that `mailpit` is properly running. On `ddev-dbserver` it will mean that the database server in the `db` container is able to respond to queries and is fully functional.
+
+Without healthchecks, containers will show up as ready when they're actually not. The `ddev-dbserver` is probably most vulnerable to this, because it can take a few seconds for the `mysqld` or `postgres` server to become ready. If we didn't wait for the `healthy` signal, traffic could try to flow before the various components were ready for it, causing ugly failures.
+
+### Avoiding CPU and battery usage
+
+There is one more key trick DDEV does with its healthchecks. During startup, we want to find out as quickly as possible when the container is healthy, so that the start can continue immediately. So we want to test quickly until it's ready. However, after that, not much goes wrong, so we have a tricky healthcheck script that, once the `web` or `db` container has become healthy, slows down the checks to about 60 seconds each. Essentially, if a healthy status has already been detected, the healthcheck script [sleeps 60 seconds](https://github.com/ddev/ddev/blob/master/containers/ddev-webserver/ddev-webserver-base-scripts/healthcheck.sh#L9-L21) to prevent another check from happening, slowing down the checking process and avoiding unnecessary use of CPU and battery.
```

Comment:
```suggestion
There is one more key trick DDEV does with its healthchecks. During startup, we want to find out as quickly as possible when the container is healthy, so that the start can continue immediately. So we want to test quickly until it's ready. However, after that, not much goes wrong, so we have a tricky healthcheck script that, once the `web` or `db` container has become healthy, slows down the checks to about 60 seconds each. Essentially, if a healthy status has already been detected, the healthcheck script [sleeps 60 seconds](https://github.com/ddev/ddev/blob/5c546c28e1c71b26b7cfb8a683e21a190a2f49dc/containers/ddev-webserver/ddev-webserver-base-scripts/healthcheck.sh#L9-L21) to prevent another check from happening, slowing down the checking process and avoiding unnecessary use of CPU and battery.
```